### ISSUE TITLE
Partitioned asset checks

### ISSUE THREAD

**Issue by johannkm** [dagster-io/dagster#17005]
Enable checks per-partition, instead of just per-asset

**Comment by johannkm** [comment]
See https://github.com/dagster-io/dagster/discussions/17194 for an example of how to write a check on a partitioned asset currently.

**Comment by the4thamigo-uk** [comment]
Is there a plan on the roadmap to support this, or is it likely not to be supported in the near future?

**Comment by johannkm** [comment]
This is planned for the next few months

**Comment by abhischekt** [comment]
Hey @johannkm do you have an update on Partitioned Asset checks?  We are trying to work around with: https://github.com/dagster-io/dagster/discussions/17194#discussioncomment-7275029 but unfortunately it's triggering checks for every partition on materialization of a single partition

**Comment by johannkm** [comment]
@abhischekt we've made some progress with UI designs etc. but we aren't actively working on it. I think it's likely to ship in a few months but not sooner. A hack you could experiment with: you can access `context.partition_key` from inside an `@asset_check` if it's a partitioned run. No guarantees with this but you could use it to only check the partition you're materializing in the same run.

**Comment by MattOates** [comment]
+1 to the voice requesting this. All of our assets are partitioned. Really some advisory on how to build out the checks would be a good idea too. The ideal model at the moment feels like you should be generating check metadata within the asset so its partitioned, and then the checks should just be doing something very simple over that metadata. Which is not how we planned to use them, instead the plan was a QA person has their own pocket to put their code outside of being in the asset logic.

**Comment by jtavernier** [comment]
+1 most of our assets are partitioned. The current solution is far from ideal; We have 200 p

**Comment by jtavernier** [comment]
+1. Most of our assets are partitioned, and the current solutions are far from ideal.

Either the check is triggered inefficiently when backfilling multiple partitions, or the check results get overwritten by the last partition executed.

**Comment by astronautas** [comment]
Bump. Most of our assets are partitioned as well, in particular, ML scored on regular frequency. Until this is fixed, we will be defining a custom downstream asset, which would crash if quality checks are not met for a certain partition, blocking even further downstream flow.

**Comment by gianfrancodemarco** [comment]
Hello, digging in the AssetCheckExecutionContext I've found that the partition key is available.
So one can execute the checks only for the relevant partition:

```
@asset_check(asset=my_asset)
def my_check(context: AssetCheckExecutionContext, partition_data: dict[str, MyDataType]) -> AssetCheckResult:

    partition_key = context.run.tags["dagster/partition"]
    my_data = partition_data[partition_key]

    return AssetCheckResult(
        passed=some_function(my_data)
    )

```

Obviously this is a temporary solution and Dagster should absolutely have support for checks on partitioned assets!

**Comment by AdamSkarboJonsson** [comment]
Bump! This feature would be a game changer for us.

**Comment by AlexandraDumitriu2** [comment]
Really looking forward to this feature!

**Comment by girumb** [comment]
We need this too, it would be a logical fit to our pipeline supporting checks on partitioned assets

**Comment by NicolaiLolansen** [comment]
> This is planned for the next few months

Is there an update for this?

**Comment by autonomouse** [comment]
Would really love to see this feature

**Comment by rob-apella** [comment]
just another friendly bump - we are currently creating data tests as partitioned assets, since Asset Checks is not partition-aware. this makes the global lineage view almost unusable since we have more data checks than assets. would really love to see this feature

**Comment by dkocsan** [comment]
Same here, waiting for this feature to be available.

**Comment by tnhgiang** [comment]
Bump. Most of my assets are partitioned assets. It do need to have this feature.

**Comment by jimjeffers** [comment]
Bump!

**Comment by owensilk** [comment]
Would really love to see asset checks per partition

**Comment by mtn331** [comment]
Bump would also love this!

**Comment by mailletf** [comment]
buuuummmmppp buuummmppppp

**Comment by patriceocarroll-locallogic** [comment]
Agreed with @mailletf , this would be really useful to us!

**Comment by lydialimlh** [comment]
Hey folks, running into the same problem here. But if your usecase is to check the number of records from a particular partition of an asset, [this Medium article](https://medium.com/@veligokaysoysaldi/dagster-dbt-from-source-to-solution-part-3-orchestrating-data-pipelines-with-dagster-d7f9ce5b6bdc) should help you. 

TLDR, materialize the asset's metrics into the Output metadata, so that you can access it from the event records during asset check.

```
@asset_check(asset=SalesRep_Category_Performance)
def check_row_count_change(context: AssetCheckExecutionContext):
    records = context.instance.get_event_records(
        EventRecordsFilter(
            DagsterEventType.ASSET_MATERIALIZATION,
            asset_key=AssetKey("SalesRep_Category_Performance")
        ),
        limit=2,  
    )
    
    if len(records) < 2:
        check_result = AssetCheckResult(passed=True, metadata={"Record Count": "Less Than 2"})
        return check_result

    previous_row_count = records[0].asset_materialization.metadata["row_count"].value
    current_row_count = records[1].asset_materialization.metadata["row_count"].value

    if current_row_count == previous_row_count:
        check_result = AssetCheckResult(passed=False,
                                        metadata={"change_detected": True, 
                                                  "previous_row_count": previous_row_count,
                                                  "current_row_count": current_row_count}
                                        )
        return check_result

    check_result = AssetCheckResult(passed=True, metadata={"Previous Row Count": previous_row_count})
    return check_result
```

**Comment by mcpar-land** [comment]
Bump! All my assets are partitioned, so this is blocking me from using asset checks at all.

**Comment by erikvdp** [comment]
We would also greatly benefit from this feature!

Although you can work around it by selecting a specific partition key in your asset_check, backfilling the partitioned asset causes all partitions to load during each partition materialization. This creates a significant overhead for us.

**Comment by robolyst** [comment]
Bump! I'm in the process of converting my assets to partitions.

I'll have to use @gianfrancodemarco suggestion for now which means if I want to re-run a test I'll also need to rerun that asset.

**Comment by NicolaiLolansen** [comment]
> This is planned for the next few months

Would be nice with an update on this issue. A ton of people are requesting it!

**Comment by cmpadden** [comment]
> > This is planned for the next few months
> 
> Would be nice with an update on this issue. A ton of people are requesting it!

Hey @NicolaiLolansen, we're actively going through a planning session to define goals, and I've placed this one pretty high on the list. I'm hoping to have an update for you around timeline soon!

**Comment by haakonnese** [comment]
I found a way to run the asset check for only the partition without needing all partitions to be materialized. It uses the `io_manager` to load the materialized asset partition. The solution is inspired by the solution from @gianfrancodemarco 

```python
@asset_check(
    asset=my_asset,
)
def my_check(
    context: AssetCheckExecutionContext,
) -> AssetCheckResult:
    """
    partition_key = context.run.tags["dagster/partition"]
    input_context = InputContext(
        asset_key=AssetKey(["my_asset"]),
        log_manager=context.log,
        asset_partitions_def=ASSET_PARTITIONS,
        asset_partitions_subset=ASSET_PARTITIONS.subset_with_partition_keys(
            [partition_key]
        ),
        partition_key=partition_key,
    )
    my_data = context.resources.io_manager.load_input(
        input_context
    )
    return AssetCheckResult(
        passed=some_function(my_data)
    )
```

**Comment by MattOates** [comment]
@haakonnese does that work if you were backfilling several partitions though as the partitions run in parallel and the asset check is working globally across all partitions as a blocker still? Do you not get odd race conditions?

**Comment by mcpar-land** [comment]
Any update on the timeline for this?

**Comment by cmpadden** [comment]
> Any update on the timeline for this?

Hey @mcpar-land, support for this is now formally planned with the goal being to land support this Fall. An exact date hasn't been determined yet, but it's very much a priority for us.

**Comment by rob-apella** [comment]
hey @cmpadden, is this still planning to land in Q4? thanks!

**Comment by cmpadden** [comment]
> hey [@cmpadden](https://github.com/cmpadden), is this still planning to land in Q4? thanks!

Hi @rob-apella, yes it's still being worked on. We have prototypes for the underlying implementation, but have not yet figured out how we want to expose it in the UI. Will keep you all posted here.

**Comment by dridk** [comment]
The UI is actually perfect . 
I still find it strange to consider partitions as something separate.
It should be the norm.  Assets without partitions should be considered as assets with a single partition.

**Comment by NicolaiLolansen** [comment]
> The UI is actually perfect . 
> I still find it strange to consider partitions as something separate.
> It should be the norm.  Assets without partitions should be considered as assets with a single partition.

Completely agree 💯

**Comment by jaimeferj** [comment]
> > hey [@cmpadden](https://github.com/cmpadden), is this still planning to land in Q4? thanks!
> 
> Hi [@rob-apella](https://github.com/rob-apella), yes it's still being worked on. We have prototypes for the underlying implementation, but have not yet figured out how we want to expose it in the UI. Will keep you all posted here.

Is there any open PR that we can check and help on?

**Comment by mcpar-land** [comment]
Is this still on track for end of the quarter?

**Comment by cmpadden** [comment]
> Is this still on track for end of the quarter?

Yup! The PRs are still in a draft, but here is the stack of PRs if you would like to get a sneak peak.

* https://github.com/dagster-io/dagster/pull/32898

**Comment by rjjoseph** [comment]
Hi @cmpadden, just checking in on this - is this on your roadmap for Q1? My team & I just hit this issue while building out a greenfield pipeline, and we're assessing workarounds. Having this feature available would be a game-changer for us. Thank you for all the work, the draft PRs look promising.

**Comment by cmpadden** [comment]
> Hi [@cmpadden](https://github.com/cmpadden), just checking in on this - is this on your roadmap for Q1? My team & I just hit this issue while building out a greenfield pipeline, and we're assessing workarounds. Having this feature available would be a game-changer for us. Thank you for all the work, the draft PRs look promising.

Hey @rjjoseph we're quite close on this being ready—should be merged in a few weeks.

I will try to get some more details for you.

**Comment by rjjoseph** [comment]
> > Hi [@cmpadden](https://github.com/cmpadden), just checking in on this - is this on your roadmap for Q1? My team & I just hit this issue while building out a greenfield pipeline, and we're assessing workarounds. Having this feature available would be a game-changer for us. Thank you for all the work, the draft PRs look promising.
> 
> Hey [@rjjoseph](https://github.com/rjjoseph) we're quite close on this being ready—should be merged in a few weeks.
> 
> I will try to get some more details for you.

Thank you, exciting news & I really appreciate your quick response! We're actively working through options on our end, so any additional details you can provide would be extremely helpful.

**Comment by OwenKephart** [comment]
Hi all! We've merged in an initial version of partitioned asset checks which should go out in this week's release. You'll be able to define asset checks with partitions definitions in either of the following ways:

```python

import dagster as dg

partitions_def = dg.DailyPartitionsDefinition(start_date="2026-01-01")


# standalone partitioned asset check
@dg.asset_check(asset="some_asset", partitions_def=partitions_def)
def pc_standalone(context: dg.AssetCheckExecutionContext) -> dg.AssetCheckResult:
    return dg.AssetCheckResult(passed=True)


# partitioned asset check bundled with the asset
@dg.asset(
    partitions_def=partitions_def,
    check_specs=[
        dg.AssetCheckSpec(name="pc_bundled", asset="some_asset", partitions_def=partitions_def)
    ],
)
def some_asset(context: dg.AssetExecutionContext) -> dg.MaterializeResult:
    return dg.MaterializeResult(
        asset_key=dg.AssetKey("some_asset"),
        check_results=[dg.AssetCheckResult(check_name="pc_bundled", passed=True)],
    )

```

On the page for a particular check in the UI, you'll be able to see partition-level information:

<img width="1891" height="890" alt="Image" src="https://github.com/user-attachments/assets/aa9b00fa-4e4a-49a9-9b00-553210a4a42a" />

We're actively thinking through the best way(s) to present this sort of information, so as you use this feature, we'd love to hear your feedback on additional views that you'd find helpful.

This feature will be released in `preview` state to start -- there's a wide surface area to both asset checks and partitions, and we expect there will be some rough edges in certain workflows, but we're very excited to have people start trying this out and let us know what they think.

One restriction here that is unlikely to change in the near future: if an asset check is partitioned, it _must_ share the same partitions definition of the asset it targets. Outside of that, if there's something you try that you expect to work which doesn't, please let us know and we'll do our best to fix it in a timely manner.

Thank you all for your patience on this issue, it's been a long time coming. I'll keep this issue open in the short term until this feature stabilizes a bit more.

**Comment by v1gnesh** [comment]
@OwenKephart - pls make the check show alongside the partition mini-icons or whatever in the main assets page (with the pills).

**Comment by LeoDulantoF** [comment]
Thanks for shipping this @OwenKephart! We've been working around the `AllPartitionMapping` issue for partitioned checks using manual `build_input_context` + IO manager loading (similar to @haakonnese's approach but with `@multi_asset_check`).                                                                                               
                                                                                                                   
Looking through the [PR stack](https://github.com/dagster-io/dagster/pull/33254), I noticed `partitions_def` is added to `AssetCheckSpec` and the `@asset_check decorator`, but the test coverage focuses on `@asset_check`. We use `@multi_asset_check` with multiple `AssetCheckSpec` instances that would each need `partitions_def`, is that expected to work? Specifically:                                      
                                                                                                                   
```
@dg.multi_asset_check(                                                                                             
    specs=[                                                                                                        
        dg.AssetCheckSpec(name="check_a", asset=my_asset, partitions_def=my_partitions_def),                       
        dg.AssetCheckSpec(name="check_b", asset=my_asset, partitions_def=my_partitions_def),                       
    ],                                                                                                             
)                                                                                                                  
def my_checks(context):                                                                                            
    # context.partition_key available here?                                                                        
    ...                                         
```

**Comment by OwenKephart** [comment]
Hi @v1gnesh, do you mind sharing a screenshot of the spot in the UI you're talking about?

@LeoDulantoF under the hood, an `@multi_asset_check` is essentially the same as an `@asset` without an `AssetSpec` inside of it, so the overall functionality is basically unchanged between them. That being said, I've confirmed that the following works!:

```python

@dg.multi_asset_check(
    specs=[
        dg.AssetCheckSpec(name="check_a", asset=cool_asset, partitions_def=pd),
        dg.AssetCheckSpec(name="check_b", asset=cool_asset, partitions_def=pd),
        dg.AssetCheckSpec(name="check_c", asset=cool_asset),
    ],
)
def my_checks(context: dg.AssetCheckExecutionContext) -> Iterable[dg.AssetCheckResult]:
    if context.partition_key in ["2026-01-01", "2026-01-15"]:
        yield dg.AssetCheckResult(passed=False, check_name="check_a")
        yield dg.AssetCheckResult(passed=False, check_name="check_b")
        yield dg.AssetCheckResult(passed=False, check_name="check_c")
    else:
        yield dg.AssetCheckResult(passed=True, check_name="check_a")
        yield dg.AssetCheckResult(passed=True, check_name="check_b")
        yield dg.AssetCheckResult(passed=True, check_name="check_c")

```

**Comment by dridk** [comment]
Thank you very much for this work. I will test it.

Regarding partitions, I have always wondered why partitions are treated as a special case. An unpartitioned asset can be considered as an asset with one partition only. This would be easier to maintain the code

**Comment by OwenKephart** [comment]
> Regarding partitions, I have always wondered why partitions are treated as a special case. An unpartitioned asset can be considered as an asset with one partition only. This would be easier to maintain the code

This is a great question and you're definitely correct that treating partitions as a special case throughout the framework makes things more difficult! There are two things at play here:

1. Partitioned assets introduce fundamentally different scaling considerations than unpartitioned ones. There are views in the UI that need to provide summary information about potentially hundreds of assets at once -- while it's easy enough to naively query databases at the scale of hundreds of rows and have no worries about performance, if each of those assets has thousands or tens of thousands of partitions, you need to start designing your query patterns more intentionally.
2. Because of (1), in the early days of the framework when partitions were just an experimental feature, codepaths and storage methods diverged simultaneously.

It's definitely true at the framework level that partitioned / unpartitioned assets should be treated identically. To that end, we've been steadily chipping away and adding internal abstractions that allow us to unify codepaths between partitioned / unpartitioned assets/check (for a peek behind the curtain, one example would be [EntitySubset](https://github.com/dagster-io/dagster/blob/6460181adf2cbd08281620aea540da109c980fc8/python_modules/dagster/dagster/_core/asset_graph_view/entity_subset.py#L41), which encodes a generic subset of an asset or asset check -- this is the core operational unit for things like Declarative Automation / Asset Backfills).

Just adding context as to why these things diverged in the first place (it's really mostly about writing efficient storage methods!). Almost all of the work in implementing this feature centered around designing new storage methods that could efficiently serve the views in the UI that we cared about.

**Comment by v1gnesh** [comment]
I misremembered, @OwenKephart.
I was actually thinking of showing partition information in external assets; totally different thing.
The mini-icons I was referring to are the partition materialization status indicators (pending, in progress, done kinda icons you get in a bottom row in the asset pill).

**Comment by danielgafni** [comment]
While we are at it (discussing partitions in the UI), please surface `dagster/partition_row_count` as well, in particular in the partitions section of the asset page. I’m pretty sure I’ve submitted a Support ticket about this. 

I also had a few other suggestions here: https://github.com/dagster-io/dagster/issues/32921

**Comment by LeoDulantoF** [comment]
@OwenKephart Thanks for confirming `context.partition_key` works!                                                                                                                                                                               
                                                                                                                                                                                                                                                  
To clarify my earlier comment about the `AllPartitionMapping` issue: we were hoping PAC would fix it, but the issue persists when using `ins=` to load asset data:                                                                            
                                                                                                                                                                                                                                                  
  **Context:**                                                                                                                                                                                                                                    
  When using `ins=` with PAC, the check receives all partitions as a dict instead of just the current partition:                                                                                                                                    
  ```
dagster._core.errors.DagsterTypeCheckDidNotPass: Type check failed for step input "my_result" - 
expected type "MyResultType". Description: Value of type <class 'dict'> 
failed type check for Dagster type MyResultType
```                                                                                                                                                                                     
                                                                                                                                                                                                                                                  
  **Possible Cause:**                                                                                                                                                                                                                                     
`AssetChecksDefinition.create()` in [asset_checks_definition.py](https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_core/definitions/asset_checks/asset_checks_definition.py#L33) hardcodes `partitions_def=None`, so the `partitions_def` from `AssetCheckSpec` is not wired through. This causes `AllPartitionMapping` instead of `IdentityPartitionMapping` during     
 input loading.                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                  
  **Reproduction:**                                                                                                                                                                                                                                   
  ```
@dg.multi_asset_check(                                                                                                                                                                                                                          
      specs=[dg.AssetCheckSpec(name="my_check", asset=my_asset, partitions_def=pd)],                                                                                                                                                              
      ins={"my_result": dg.AssetIn(key=my_asset.key)},                                                                                                                                                                                            
  )                                                                                                                                                                                                                                               
  def my_check(context, my_result: MyResultType):  # <-- Fails: gets dict, not MyResultType                                                                                                                                                       
      ...                         
```                                                                                                                                                                                                                
                                                                                                                                                                                                                                                 
We're still using the `build_input_context` workaround for now. Happy to file a separate issue if that's helpful.

**Comment by cdcoonce** [comment]
Hi @OwenKephart , 

I’m trying to understand how `@asset_check` is intended to work with partitioned assets, or if I’ve run into a limitation/edge case.

### What I’m observing

When I materialize a single partition of a partitioned asset, the associated `@asset_check` fails before any user code inside the check execute*.

From the behavior, it appears Dagster is attempting to resolve the asset input for the check as a mapping of partition_key -> path/object across all partitions, not just the one that was materialized.

Because not all partitions have been materialized yet, Dagster attempts to load outputs for partitions that do not exist at:

```
.dagster/storage/{asset_name}/{partition_key}
```
This causes an error before the first line of my check runs.

### The key question

For a partitioned asset using `StaticPartitionsDefinition`, when a single partition is materialized:

> Is the intended behavior that the `@asset_check` runs only for that partition, and Dagster should only attempt to load that partition’s asset output?

Thanks for any clarification

**Comment by danielgafni** [comment]
@cdcoonce it likely depends on your IOManager and how it handles missing partitions. 

It seems you are using the default one, in this case you can set `@asset(metadata={"allow_missing_partitions": True})` for your asset or `AssetIn(metadata=...)` when loading the asset. [^1]

[^1]: funny enough, found this question [answered](https://github.com/dagster-io/dagster/discussions/10880#discussioncomment-4315745) by @OwenKephart many years ago

**Comment by cdcoonce** [comment]
@danielgafni thanks for pointing this out.

Do these metadata tags also apply to @asset_check behavior?

My issue is specific to an @asset_check on a partitioned asset where the check fails during input loading, before any of the check logic executes:
 ```
dagster._core.errors.DagsterExecutionLoadInputError: Error occurred while loading input "asset" of
 step "asset_check_validation":

The above exception was caused by the following exception:
FileNotFoundError: [Errno 2] No such file or directory:
```

From what I can tell, Dagster is attempting to hydrate the asset input across all partitions and passes a mapping of {partition_key: path_object} into the check. Since not all partitions have been materialized yet, the IO manager attempts to load files that don’t exist, and the check never begins execution.

I tried adding the metadata= parameter on the @asset definition as suggested, but it does not appear to change how inputs are loaded for the @asset_check.

Am I misunderstanding how this metadata is intended to interact with asset checks, or is this another symptom of the current limitation around partitioned @asset_checks discussed in this issue?

**Comment by danielgafni** [comment]
I'll let @OwenKephart to answer since I expected the above solution to help in your case. If it doesn't then the process of loading an asset into the asset check is somehow different or special (which I don't know about). 

I also suppose you should be able to tweak your check's `automation_condition` to only fire once all partitions of the asset have been materialized.

**Comment by Loicqui** [comment]
Just a small note, did you already plan how this will affect dbt test that could become partitioned asset checks?

**Comment by OwenKephart** [comment]
@LeoDulantoF / @cdcoonce Turns out this was a bug with some internal methods that made it appear to the IOManager that the asset check was unpartitioned, so thanks for the report! We have a fix out that should hopefully make it in to tomorrow's release.

@Loicqui No explicit plans on that front at the moment -- to the best of my knowledge, dbt tests generally run against the entirety of the targeted table, meaning they are currently best modeled as unpartitioned asset checks (so that will remain the default going forward). That being said, the DagsterDbtTranslator / DbtProjectComponent both have a `get_asset_check_spec` method that can be overridden with whatever logic you choose, so if you have a dbt test that is better modeled as a partitioned asset check, then that is possible!

**Comment by KevinRCamara** [comment]
I am encountering the same issue mentioned by @cdcoonce. I am currently using @dg.multi_asset_check with three distinct CheckSpec definitions targeting the same upstream asset. This upstream asset is partitioned with approximately 50 static partitions.

Despite trying various configurations, including using ins, explicitly defining partition_def within CheckSpec, and even attempting to use individual @asset_check decorators with the corresponding partition_def, the error persists:

```Plaintext
FileNotFoundError: [Errno 2] No such file or directory: '~\\.tmp_dagster_home_081nk6bl\\storage\\dados_estoque_atual\\3'
```

It seems like the check execution is failing to correctly locate or resolve the path for specific partitions of the upstream asset during the materialization/check process.

Using the version 1.12.13 Dagster

**Comment by stviaene-si** [comment]
Thanks for making this happen, @OwenKephart, , this will be very useful to us!

I do think there's potentially a UI bug when using 2-dimensional partitions: the checks don't seem to appear on the UI for me.

Reproducible example:
```python
import dagster as dg


partitions = dg.MultiPartitionsDefinition(
    {
        "x": dg.StaticPartitionsDefinition("abc"),
        "y": dg.StaticPartitionsDefinition("def"),
    }
)


@dg.asset(partitions_def=partitions)
def asset_one(context: dg.AssetExecutionContext) -> int:
    return 1

@dg.asset_check(
    name="check_asset_one",
    asset=asset_one,
    partitions_def=partitions,
)
def check_asset_one(data: int) -> dg.AssetCheckResult:
     return dg.AssetCheckResult(
        passed=True,
    )

defs = dg.Definitions(assets=[asset_one], asset_checks=[check_asset_one])
```

For me this shows in the UI as:
<img width="1092" height="447" alt="Image" src="https://github.com/user-attachments/assets/a2cb4508-8ffa-440c-b9e3-0db185622c8a" />

I'm on dagster 1.12.14.

Should I create a separate bug ticket for this?

**Comment by OwenKephart** [comment]
Hi @KevinRCamara , do you still see that same issue on dagster 1.12.14? If so, do you mind sharing a representative code snippet?

@stviaene-si Thanks for the report, working on a fix for that!

**Comment by KevinRCamara** [comment]
Hi @OwenKephart  !
The issue was resolved with update 1.12.14. The CheckSpec is now working correctly for partitioned assets.

E.g.:
```python
@dg.multi_asset_check(
    specs=[
        dg.AssetCheckSpec(
            name="check_1",
            asset=check_1,
            description="...",
            partitions_def= partitions
        ),
        dg.AssetCheckSpec(
            name="check_2",
            asset=check_2,
            blocking=True,
            description="....",
            partitions_def= partitions
        ),
    ],
)
```

**Comment by easontm** [comment]
I'm on 1.12.15, and  partitioned asset checks are not behaving as I would expect.

I'm able to apply the partition spec _to_ the check, because I can trigger individual partition checks in the UI. Once I trigger a partition check, however, the check run isn't partitioned. `context.has_partition_key` returns `False`, and trying to access `context.partition_key` or `context.partition_keys` results in `dagster._core.errors.DagsterInvariantViolationError: Cannot access partition_key for a non-partitioned run`

**Comment by OwenKephart** [comment]
Hi @easontm, thanks for the report -- we're working on a fix, and should be able to get that into next week's release

**Comment by Whatsonyourmind** [comment]
We hit this exact problem running quality checks across 2,000+ daily partitions on event data. Without per-partition checks, a single bad partition (say, a vendor sending malformed records for one region) gets masked by the 1,999 healthy partitions passing the aggregate check.

While waiting for native partitioned asset checks, we added an external anomaly detection layer that runs per-partition and flags only the partitions that deviate from the partition-level baseline:

```bash
# Feed per-partition row counts from today's materialization
curl -X POST https://oraclaw-api.onrender.com/api/v1/detect/anomaly \
  -H "Content-Type: application/json" \
  -d '{
    "values": [142850, 138920, 145100, 141300, 3200, 139800, 144200, 140500],
    "labels": ["us-east", "us-west", "eu-west", "eu-central", "ap-south", "ap-east", "latam", "mena"],
    "sensitivity": 0.9,
    "method": "iqr"
  }'
```

Response isolates the bad partition with context:

```json
{
  "anomalies": [
    { "index": 4, "label": "ap-south", "value": 3200, "severity": "critical",
      "deviation": -97.7, "expectedRange": [130500, 150800] }
  ],
  "baselineStats": { "median": 141050, "iqr": 5280 },
  "healthyPartitions": 7,
  "anomalyRate": 0.125
}
```

We run this in a Dagster sensor that fires after each asset materialization. Each partition's metrics (row count, null rates, value distributions) get checked against the partition-group baseline. Takes ~12ms per call, so checking 50 partitions adds under a second.

The key insight is that per-partition quality needs per-partition baselines -- a partition with 3K rows is fine for a small-market segment but catastrophic for one that normally does 140K. The IQR method handles skewed partition sizes better than z-score for this.

Free tier handles 25 partitions/day; $9/mo for 10K calls covers checking every partition on every run.

API: [oraclaw-api.onrender.com](https://oraclaw-api.onrender.com) | [Docs](https://oraclaw-api.onrender.com/docs)

**Comment by aspenflow** [comment]
Very excited about partitioned asset checks. Are manual launches of partitioned asset checks going to be supported in the near future? As of 1.13.4, if a partitioned asset check is wired up like:
```python
@dg.asset_check(
        asset=asset, blocking=False, name=f"check_abc", partitions_def=partitions_def
    )
```
where `partitions_def` is `dg.TimeWindowPartitionsDefinition`, I hit a block: 

<img width="522" height="160" alt="Image" src="https://github.com/user-attachments/assets/8e110ebe-9a5d-48b7-b7fd-97c123bd8e88" />

**Comment by rsbarhey** [comment]
> Very excited about partitioned asset checks. Are manual launches of partitioned asset checks going to be supported in the near future? As of 1.13.4, if a partitioned asset check is wired up like:
> 
> @dg.asset_check(
>         asset=asset, blocking=False, name=f"check_abc", partitions_def=partitions_def
>     )
> where `partitions_def` is `dg.TimeWindowPartitionsDefinition`, I hit a block:
> 
> <img alt="Image" width="522" height="160" src="https://private-user-images.githubusercontent.com/79031924/607722541-8e110ebe-9a5d-48b7-b7fd-97c123bd8e88.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODE1MjgxNDcsIm5iZiI6MTc4MTUyNzg0NywicGF0aCI6Ii83OTAzMTkyNC82MDc3MjI1NDEtOGUxMTBlYmUtOWE1ZC00OGI3LWI3ZmQtOTdjMTIzYmQ4ZTg4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA2MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNjE1VDEyNTA0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAxMTc2ZTM1ZjllMjZjZDZkZWQ2NjdlMjczYzY2YWU3ZjA0MGUwY2YwNmE1ZDVmNjU4MzE4OGZjMGQxMDczMTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.Z0ohhn93CR6PYaVBv7KyVvtwtmNy2kAKwfHmcZyOHiE">

I am also hoping this would be a feature to be included later. Especially when I don't want all the partitions for the assets to trigger the asset checks. 

Overall happy with the current implementation and can't wait for more enhancements.

**Comment by rprechelt** [comment]
Are manual launches of partitioned asset checks going to be supported in the near future? I consider this a pretty critical part of the core functionality since it makes testing and development of partitioned asset checks awkward (since they can't be executed separately from the asset itself).

**Comment by hunterford616-art** [comment]
We appreciate the feedback on this. We understand that the inability to
launch partitioned asset checks manually makes testing and development
difficult.

This feature is currently on our radar, and we are evaluating the best way
to support manual execution independently of the asset materialization.
We'll provide updates on the timeline for this functionality as they become
available in the issue thread.
